src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -69,6 +72,23 @@ public VisitResult(TypeWithState rValueType, TypeWithAnnotations lValueType)
                 $"RValueType={RValueType.GetDebuggerDisplay()}, LValueType={LValueType.GetDebuggerDisplay()}";
         }
 
+        /// <summary>
+        /// Represents the result of visiting an argument expression.
+        /// In addition to storing the <see cref="VisitResult"/>, also stores the <see cref="LocalState"/>
+        /// for reanalyzing a lambda.
+        /// </summary>
+        [DebuggerDisplay("{GetDebuggerDisplay(), nq}")]
+        private readonly struct VisitArgumentResult
+        {
+            public readonly VisitResult VisitResult;
+            public readonly Optional<LocalState> StateForLambda;
```

Comment:
>Optional<LocalState> [](start = 28, length = 20)

Why `Optional<>` rather than `Nullable<>`?

Reply:
Presumably because `Optional<T>` does not have a `struct` constraint on `T`.

---
In reply to: [267573904](https://github.com/dotnet/roslyn/pull/34300#discussion_r267573904) [](ancestors = 267573904)

Reply:
Yes. `LocalState` is defined like this:

``` c#
#if REFERENCE_STATE
        internal class LocalState : ILocalState
#else
        internal struct LocalState : ILocalState
#endif
```

where

``` c#
#if DEBUG
// See comment in DefiniteAssignment.
#define REFERENCE_STATE
#endif
```

---
In reply to: [267576403](https://github.com/dotnet/roslyn/pull/34300#discussion_r267576403) [](ancestors = 267576403,267573904)